### ISSUE TITLE
Investigate random iconv SEGFAULTs in the CI runner

### ISSUE THREAD

**Issue by Duckwhale** [evo-lua/evo-runtime#300]
I probably did something stupid, but it's unclear why this never happens on my machines and only occasionally during CI runs.

Example workflow: https://github.com/evo-lua/evo-runtime/actions/runs/6332991564/job/17200414180

```sh
Run evo Tests/unit-test.lua
hi: 0 ms
test: 234 ms
WARNING: iconv_open failed with error code 92 (Illegal byte sequence)
Input: 유저인터페이스
Input Encoding: CP949
Output Encoding: UTF-8
WARNING: iconv_open failed with error code 22 (Invalid argument)
Input: 유저인터페이스
Input Encoding: INVALID_ENCODING
Output Encoding: UTF-8
WARNING: iconv_open failed with error code 22 (Invalid argument)
Input: 유저인터페이스
Input Encoding: UTF-8
Output Encoding: INVALID_ENCODING
/Users/runner/work/_temp/da3afe20-7e47-4327-8053-b4de2b8a49ea.sh: line 1: 28610 Segmentation fault: 11  evo Tests/unit-test.lua
```

~~TBD: Is this affecting OSX only?~~ Nope, also fails in the Windows CI (sometimes).

**Comment by Duckwhale** [comment]
Since I can't reproduce the problem locally, I'll just assume the changes made in #307 fix it. If not, reopen and investigate more.

**Comment by Duckwhale** [comment]
Seems to still be a thing, but only in the CI. It's not even clear if this is from iconv, could (in theory) also be elsewhere.

For now, I'll add some more debug output - the entire iconv glue probably needs a rework anyway, which I'll tackle later.

**Comment by Duckwhale** [comment]
Managed to get it a few times, but it's extremely sporadic. Can take hundreds or even thousands of runs to crash locally...

From what I can gather, the issue lies in the [large input test](https://github.com/evo-lua/evo-runtime/blob/9eb62cafeaaf23e25dcf5997657aa45a8772261e/Tests/BDD/iconv-library.spec.lua#L90-L93):

```sh
invalid input encoding test done
[bdd] Reporting message:     [0;32m✓[0;0m should fail gracefully if given an invalid input encoding
WARNING: iconv_open failed with error code 22 (Invalid argument)
Input: 유저인터페이스
Input Encoding: UTF-8
Output Encoding: INVALID_ENCODING
invalid output encoding test done
[bdd] Reporting message:     [0;32m✓[0;0m should fail gracefully if given an invalid output encoding
num_processed_bytes: 1000000
```

This output suggests something went wrong when the large string is supposed to be GCed:

```sh
Output Encoding: INVALID_ENCODING
AddressSanitizer:DEADLYSIGNAL
=================================================================
==104999==ERROR: AddressSanitizer: SEGV on unknown address 0x7f9492690b16 (pc 0x55c071fa9e55 bp 0x000000000000 sp 0x7fff528e8c70 T0)
==104999==The signal is caused by a READ memory access.
    #0 0x55c071fa9e55 in lj_alloc_free (evo+0x57ce55) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #1 0x55c071fb3078 in gc_sweepstr (evo+0x586078) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #2 0x55c071fb3e0f in gc_onestep (evo+0x586e0f) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #3 0x55c071fb42fc in lj_gc_step (evo+0x5872fc) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #4 0x55c071fa0f07 in lj_parse_keepstr (evo+0x573f07) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #5 0x55c071f9e82c in lex_scan (evo+0x57182c) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #6 0x55c071f9ea20 in lj_lex_next (evo+0x571a20) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #7 0x55c071fa2a67 in parse_args (evo+0x575a67) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #8 0x55c071fa2b95 in expr_primary (evo+0x575b95) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #9 0x55c071fa315a in parse_chunk (evo+0x57615a) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #10 0x55c071fa10c4 in parse_body (evo+0x5740c4) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #11 0x55c071fa204a in expr_binop (evo+0x57504a) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #12 0x55c071fa2a7f in parse_args (evo+0x575a7f) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #13 0x55c071fa2b95 in expr_primary (evo+0x575b95) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #14 0x55c071fa315a in parse_chunk (evo+0x57615a) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #15 0x55c071fa10c4 in parse_body (evo+0x5740c4) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #16 0x55c071fa204a in expr_binop (evo+0x57504a) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #17 0x55c071fa2a7f in parse_args (evo+0x575a7f) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #18 0x55c071fa2b95 in expr_primary (evo+0x575b95) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #19 0x55c071fa315a in parse_chunk (evo+0x57615a) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #20 0x55c071fa47bc in lj_parse (evo+0x5777bc) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #21 0x55c071fa4889 in cpparser (evo+0x577889) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #22 0x55c071fb0bf0 in lj_vm_cpcall (evo+0x583bf0) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #23 0x55c071fa49e4 in lua_loadx (evo+0x5779e4) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #24 0x55c071fa4ac2 in luaL_loadfilex (evo+0x577ac2) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #25 0x55c071fdf8de in lj_cf_dofile (evo+0x5b28de) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #26 0x55c071fb07fa in lj_BC_FUNCC (evo+0x5837fa) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #27 0x55c071fdf8f1 in lj_cf_dofile (evo+0x5b28f1) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #28 0x55c071fb07fa in lj_BC_FUNCC (evo+0x5837fa) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #29 0x55c071f9ca41 in lua_pcall (evo+0x56fa41) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)
    #30 0x55c071f85c45 in LuaVirtualMachine::RunCompiledChunk() Runtime/LuaVirtualMachine.cpp:96
    #31 0x55c071f868ba in LuaVirtualMachine::DoString(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) Runtime/LuaVirtualMachine.cpp:71
    #32 0x55c071e06043 in main Runtime/main.cpp:59
    #33 0x7f94a56d7082 in __libc_start_main ../csu/libc-start.c:308
    #34 0x55c071e1ab4d in _start (evo+0x3edb4d) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca)

AddressSanitizer can not provide additional info.
SUMMARY: AddressSanitizer: SEGV (evo+0x57ce55) (BuildId: bf1e8ca0b5a67c545d5164a34a6a75e956ec51ca) in lj_alloc_free
==104999==ABORTING
```

**Comment by Duckwhale** [comment]
Since the test is somewhat artificial and unlikely to occur in practice, I will simply remove it for the time being. The assumption being that if large strings need to be processed, then efficiency will matter and so using the "convenience" Lua API would be the wrong choice, versus using the FFI bindings directly. Should the crashes still occur after that, I'll re-open the investigation.

**Comment by Duckwhale** [comment]
Nope, it's still there:

![image](https://github.com/evo-lua/evo-runtime/assets/16489133/bc7148ab-b9b4-4cae-983f-f75b665ad430)

Since the issue occurs inside LuaJIT code, need to find a way to reproduce it. Clearly it wasn't the large string buffer after all...

**Comment by Duckwhale** [comment]
Ignoring the miniz crash for now (if it still happens later, move to a separate issue). On native Linux I can now observe the crash:

```lua
			local input = ""
			local output, numBytesWritten = iconv.convert(input, "CP949", "UTF-8")

			assertEquals(output, "")
			assertEquals(numBytesWritten, 0)
```

This is the isolated test case; all others are working. I think here is what's happening, as far as I can observe:

1.  Test case invokes `iconv.convert(input, "CP949", "UTF-8")` with the empty string
2. However, `inputBuffer` is a zero-length char buffer because the input was empty
3.  LuaJIT returns `NULL` after trying to reserve the zero-length `outputBuffer` (expected?)
4. The early exit condition in `iconv_convert` (C++) catches this and returns `0` immediately (no writes, OK)
5. The FFI bindings commit zero bytes, which is a NO-OP (OK), but the pointer is still NULL (oops)
6. The return value is correctly created, i.e. the bindings return an empty string and zero bytes written (test passes)
7. However, when LuaJIT tries to collect the NULL pointer (buffer) it seems to sometimes  fail - not sure why it isn't 100%

In short: The test case was supposed to find the problem, but it doesn't work reliably because the issue appeared later.
